Add tags to anonymous structs and unions (makes FFI bindings easier)

yallop opened the pull request:
This small PR removes an obstacle to creating bindings to gRPC, by adding tags to types that are currently anonymous. It fixes #12075, where the problem is described in more detail.

The patch adds tags to all the completely-anonymous types that form part of the public interface. Types with `typedef` aliases, but without tags, are left unchanged, as are types with only reserved members. The tag names are derived from the enclosing type and the name of the member — for example, the type of the `pointer` member in `grpc_arg` becomes struct `grpc_arg_pointer`.

grpc-testing commented:
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok for us to run on our infrastructure.

grpc-jenkins commented:
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok for us to run on our infrastructure.

grpc-jenkins commented:
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok for us to run on our infrastructure.

grpc-testing commented:
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok for us to run on our infrastructure.

grpc-jenkins commented:
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok for us to run on our infrastructure.

grpc-testing commented:
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok for us to run on our infrastructure.

grpc-jenkins commented:
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok for us to run on our infrastructure.

grpc-testing commented:
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok for us to run on our infrastructure.

grpc-jenkins commented:
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok for us to run on our infrastructure.

grpc-jenkins commented:
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok for us to run on our infrastructure.

grpc-jenkins commented:
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok for us to run on our infrastructure.

grpc-jenkins commented:
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok for us to run on our infrastructure.

grpc-jenkins commented:
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok for us to run on our infrastructure.

grpc-testing commented:
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok for us to run on our infrastructure.

grpc-jenkins commented:
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok for us to run on our infrastructure.

grpc-jenkins commented:
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok for us to run on our infrastructure.

grpc-testing commented:
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok for us to run on our infrastructure.

yallop commented:
@dgquintas, @markdroth: are you able to take a look?  I think this one should be fairly uncontroversial.

vjpai commented:
Jenkins, this is ok to test

vjpai commented:
@yallop : can you also create a matching issue describing this problem? I think most of the PR description above can be moved to the issue .

yallop commented:
@vjpai: I've opened #12075.

vjpai commented:
![Pony expressing approval](http://i0.kym-cdn.com/photos/images/original/000/284/941/1a2.png)

grpc-jenkins commented:
```
[trickle] No significant performance differences
```

grpc-jenkins commented:
```
[microbenchmarks] No significant performance differences
```

vjpai commented:
Now, interestingly, there are about 6 other anonymous structs/unions in gRPC core headers besides the ones in this pull request. Do you not need names on those?

yallop commented:
If I'm not mistaken, the remaining ones are either internal-only structs, like this

```C
struct {
  void *obfuscated[4];
} internal_data;
```

or structs with a `typedef` name but no tag, like this:

```C
typedef struct {
  size_t num_args;
  grpc_arg *args;
} grpc_channel_args;
```

The `typedef` names are sufficient for my purposes, since they can be used to retrieve layout information in just the same way as a tag:

```C
… offsetof(grpc_channel_args, args) …
```

vjpai commented:
I'm thinking, for example, of the `default_level` and `default_algorithm` fields of `compression_types.h` ; these are exposed to language wrappers (and are used in other language bindings like Ruby) but don't have a type name.

yallop commented:
@vjpai: got it; thanks!  I've added tags for the remaining types (72b8ed4) and rebased against master.

I can add a sanity check to the branch, too (https://github.com/grpc/grpc/issues/12075#issuecomment-320480517), if someone directs me to the appropriate place to add the code.

vjpai commented:
I wrote a sanity check in #12095, but it isn't working for some reason that I don't understand, so I'm going to try to get that solved so that we can merge this together with that.

grpc-jenkins commented:
```
[trickle] No significant performance differences
```

grpc-jenkins commented:
```
[microbenchmarks] No significant performance differences
```

yallop commented:
@vjpai: I've cherry-picked your sanity check from #12095 (10a6e611f4).

vjpai commented:
Oh that's funny, I just sent it as a sub-PR as well. Ok, feel free to close
that sub-PR.


On Wed, Aug 9, 2017 at 7:33 AM yallop <notifications@github.com> wrote:

> @vjpai <https://github.com/vjpai>: I've cherry-picked your sanity check
> from #12095 <https://github.com/grpc/grpc/pull/12095> (10a6e61
> <https://github.com/grpc/grpc/commit/10a6e611f42a4d49e07e85ffed2b4bfe4c15b77e>
> ).
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/grpc/grpc/pull/12059#issuecomment-321273501>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AIdEkMQ7iPRjIBmY9QMIdxkmJV81Bb0vks5sWcMogaJpZM4OsjEj>
> .
>

vjpai commented:
Requesting a second code review from @markdroth . This includes a working sanity check

markdroth commented on include/grpc/impl/codegen/byte_buffer_reader.h:
```diff
@@ -28,8 +28,8 @@ struct grpc_byte_buffer;
   struct grpc_byte_buffer *buffer_out;
-  /** Different current objects correspond to different types of byte buffers */
-  union {
+  /** Different current objects correspond to different types of byte buffers */ 
```
Nit: Please remove trailing whitespace.  (I suspect the clang-format sanity check will fail because of this.)

yallop replied:
@markdroth: fixed (027a52a).  Let me know if you'd like any of this squashed.

vjpai commented:
Yes, please squash now that you have 2 approvals.

yallop commented:
@vjpai: done.  To preserve authorship data I've left the sanity check as a separate commit.

grpc-jenkins commented:
```
[trickle] No significant performance differences
```

grpc-jenkins commented:
```
[microbenchmarks] No significant performance differences
```

vjpai commented:
Jenkins test failure: #11737 in ASAN (unrelated)

vjpai commented:
Thanks again for your contribution and please keep us posted about your OCaml bindings - it would be awesome to see those posted on grpc-io mailing list.
